### PR TITLE
Adding the squared L2 norm operator for L2 regularization

### PR DESCRIPTION
**Pull request by abhinavarora**

### CODE REVIEW

File: paddle/operators/squared_l2_norm_op.h [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  void Compute(const framework::ExecutionContext &context) const override {
+    const framework::Tensor *X = context.Input<framework::Tensor>("X");
+    framework::Tensor *Out = context.Output<framework::Tensor>("Out");
+    Out->mutable_data<T>(context.GetPlace());
+
+    auto x = framework::EigenVector<T>::Flatten(*X);
+    auto out = framework::EigenVector<T>::Flatten(*Out);
+    auto place = context.GetEigenDevice<Place>();
+
+    out.device(place) = x.square().sum();
```
Comment (qingqing01) [inline paddle/operators/squared_l2_norm_op.h]: 1.  The `SquaredL2Norm` is good name, the square of L2 norm.

2. But if for L2 regularization, usually the overall cost doest not contain the regularization term (also called a weight decay term) in most framework, including the old PaddlePaddle framework.  So there is no need to use this op in the forward. Although, the cost function (as follows) contains this term.

    see the formula in this link : http://ufldl.stanford.edu/wiki/index.php/Backpropagation_Algorithm

    And in the backward backpropagation,  [the derivative of the overall cost function  J(w, b)](http://ufldl.stanford.edu/wiki/index.php/Backpropagation_Algorithm) is: 

    see the formula in this link : http://ufldl.stanford.edu/wiki/index.php/Backpropagation_Algorithm

    the weight decay term will be a linear operation on W.  (Discussed with @lcy-seso serveral days ago).  So only needs a scale op for the L2 Regularization in the parameter updating process.  Also can see the momentum updating formula in the [paper](http://papers.nips.cc/paper/4824-imagenet-classification-with-deep-convolutional-neural-networks), the weight decay is a linear operation on W.

So I'm not sure whether this op is needed and whether there is another use scenarios.

Comment (reyoung) [reply paddle/operators/squared_l2_norm_op.h]: 1. Regularization Loss is a term of the overall loss equation. If we want to plot an overall loss graph along every iteration for a classification task, the loss should equal `classification loss + regularization loss`.

2. `weight decay` just fits `L2-norm regularization`. We try to implement a common way for regularization, which fits both L1 and L2. 

3. We will provide a `weight decay` in the future PRs. It is not a conflict that we provide both `weight decay` and `L2 Norm` operator.

Comment (abhinavarora) [reply paddle/operators/squared_l2_norm_op.h]: @qingqing01 @lcy-seso Both these regularization techniques have their advantages and disadvantages. Let me summarize them as follows:
 
| Weight Decay in Optimizer 	| Separate Operators for regularization 	|
|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------	|----------------------------------------------------------------------------------------------------------:	|
| No Forward prop op, hence faster 	| There will be a forward prop op 	|
| Will not support making plots of total loss vs epoch/iteration 	| Will support those plots 	|
| This does not generalize well beyond L2 and L1 regularization. Use of Batch norm and layer norm might invalidate this approach 	| This is a very general approach and can be applied to any  kind of network 	|
| It is not easy for researchers to add new regularizers in this framework because the regularization is tightly coupled with optimizers. They might  have to change all optimizers. 	| Adding new regularization schemes is easy as the code for regularization  is independent of optimization. 	|
| Frameworks that support this: Pytorch, Caffe 	| Frameworks that support this: Tensorflow, Theano, Lasagne 	|

Comment (abhinavarora) [reply paddle/operators/squared_l2_norm_op.h]: I think that forward prop op might be necessary because while training the model, it is a very common practice to check convergence by making plots of the total loss function vs time. Also during inference, an intelligent executor can easily prune out the graph to remove the regularization nodes. 

Also I agree with @reyoung that we can also implement the weight decay separately. That can be only for the case of the L2 and L1 Penalty loss. 

Please let me know what you think about this plan?

File: paddle/operators/squared_l2_norm_op.cc [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+class SquaredL2NormOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+  void InferShape(framework::InferShapeContext* ctx) const override {
+    PADDLE_ENFORCE(ctx->HasInput("X"), "Input(X) should be not null.");
+    PADDLE_ENFORCE(ctx->HasOutput("Out"), "Output(Out) should be not null.");
+
+    ctx->SetOutputDim("Out", {1});
+    ctx->ShareLoD("X", /*->*/ "Out");
```
[the remark this answers is not in the document]
Comment (abhinavarora) [reply paddle/operators/squared_l2_norm_op.cc]: Fixed in 1ff3d8dd5d6d3af402c47d29849e19a4ed636bac

File: paddle/operators/squared_l2_norm_op.cc [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  }
+};
+
+class SquaredL2NormOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  SquaredL2NormOpMaker(framework::OpProto* proto,
+                       framework::OpAttrChecker* op_checker)
+      : framework::OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("X", "(Tensor) The input of squared_l2_norm op.");
+    AddOutput("Out", "(Tensor) The output of squared_l2_norm op.");
```
[the remark this answers is not in the document]
Comment (abhinavarora) [reply paddle/operators/squared_l2_norm_op.cc]: Thank you for the feedback. Fixed in 1ff3d8dd5d6d3af402c47d29849e19a4ed636bac

File: paddle/operators/squared_l2_norm_op.h (hunk 1 above)
Comment (qingqing01) [reply paddle/operators/squared_l2_norm_op.h]: - For the plotting:
In the old PaddlePaddle framework and Caffe, I think we usually plot the loss with regularization term vs time.

- For the regularization

I always agree to separate operators for regularization, not implementing in the optimizer.  But I think whether to add the regularization loss to the overall loss depends on the users, not by default. Since it all add more calculation during the training. And the default regularization is only to use `L2LrRegularizer` (a linear operator)/ `L1LrRegularizer` operators like the implementation in https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/parameter/Regularizer.h for the parameters updating. 

- About this PR
I agree to merge this Op. But whether to use it in the forward and add the regularization loss to the overall loss depends on the users, not default by the framework, if using the regularization in the training.

Comment (abhinavarora) [reply paddle/operators/squared_l2_norm_op.h]: @qingqing01 @lcy-seso Thank you for going through the PR. I discussed this today with @reyoung and we feel that your point is valid. We can add these ops but whether to use them in forward and add to the loss, will be the user's choice. In another PR, I will add separate operators for regularization which will be used only in the backward pass and not implemented in the optimizer. 
In case of L2, this will be a simple scale op and in the case of L1 regularization this will be a combination of scale and sign op. 
Thank you so much for your feedback on this.

**Comment by qingqing01** [comment]
@abhinavarora @reyoung  I agree to merge this PR.
